class-auth.php

```diff
@@ -528,98 +613,230 @@ public function validate_token( $return_response = true ) {
 	 * @param WP_REST_Request $request The request.
 	 * @return WP_REST_Response Returns WP_REST_Response.
 	 */
-	public function refresh_token( WP_REST_Request $request ) {
-		if ( ! isset( $_COOKIE['refresh_token'] ) ) {
+	public function refresh_token( \WP_REST_Request $request ) {
+
+		$input_refresh_token = $this->retrieve_refresh_token( $request );
+
+		if ( empty( $input_refresh_token ) ) {
 			return new WP_REST_Response(
 				array(
 					'success'    => false,
 					'statusCode' => 401,
-					'code'       => 'jwt_auth_no_auth_cookie',
-					'message'    => __( 'Refresh token cookie not found.', 'jwt-auth' ),
+					'code'       => 'jwt_auth_no_refresh_token',
+					'message'    => __( 'Refresh token not found.', 'jwt-auth' ),
 				),
 				401
 			);
 		}
-		$device = $request->get_param( 'device' ) ?: '';
 
-		$user_id = $this->validate_refresh_token( $_COOKIE['refresh_token'], $device );
-		if ( $user_id instanceof WP_REST_Response ) {
-			return $user_id;
+		$payload = $this->validate_refresh_token( $input_refresh_token, false );
+		if ( $payload instanceof WP_REST_Response ) {
+			return $payload;
 		}
 
 		// Generate a new access token.
-		$user = get_user_by( 'id', $user_id );
-		$this->send_refresh_token( $user, $request );
+		$user          = get_user_by( 'id', $payload->data->user->id );
+		$device        = $payload->data->device;
+		$refresh_token = $this->send_refresh_token( $user, $device );
+
+		$additional_fields = array();
+
+		if ( ! $this->is_flow( 'cookie' ) ) {
+			$additional_fields = array(
+				'data' => array(
+					'refresh_token' => $refresh_token,
+				),
+			);
+		}
 
 		$response = array(
 			'success'    => true,
 			'statusCode' => 200,
 			'code'       => 'jwt_auth_valid_token',
 			'message'    => __( 'Token is valid', 'jwt-auth' ),
 		);
-		return new WP_REST_Response( $response );
+
+		return new WP_REST_Response( array_merge( $response, $additional_fields ) );
 	}
 
 	/**
 	 * Validates refresh token.
 	 *
-	 * @param string $refresh_token_cookie The refresh token to validate.
-	 * @param string $device The device of the refresh token.
-	 * @return int|WP_REST_Response Returns user ID if valid or WP_REST_Response on error.
+	 * @param string $refresh_token The refresh token.
+	 * @param bool $return_response Either to return full WP_REST_Response or to return the payload only.
+	 *
+	 * @return \stdClass|WP_REST_Response Returns user ID if valid or WP_REST_Response on error.
 	 */
-	public function validate_refresh_token( $refresh_token_cookie, $device ) {
-		$parts = explode( '.', $refresh_token_cookie );
-		if ( count( $parts ) !== 2 || empty( intval( $parts[0] ) ) || empty( $parts[1] ) ) {
+	public function validate_refresh_token( $refresh_token, $return_response = true ) {
```

Comment:
Did you look into these points?  Let me know if I can support.

Reply:
1. Do you think the effort to create a backward code is worth it? If yes, how could be painless the backward implementation?
2. No, the method for verification of refresh token is not exposed as API Rest. A common method that sorts the functionality and the different error code keys can be done. IMHO this will not reduce the the amount of changes but will increase it.